bsplines with elasticity inner product crash

This is on `9fe93e2af1b2b3eab509ef1d44a1fdbefbf4d90b`

```
test_levelset.py:166: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/wechsung/Documents/Uni/DPhil/shapelib/fireshape/control.py:239: in __init__
    self.inner_product.get_impl(self.V_r, self.FullIFW)
/home/wechsung/Documents/Uni/DPhil/shapelib/fireshape/innerproduct.py:69: in get_impl
    a = self.get_weak_form(V)
/home/wechsung/Documents/Uni/DPhil/shapelib/fireshape/innerproduct.py:190: in get_weak_form
    mu = self.get_mu(V)
/home/wechsung/Documents/Uni/DPhil/shapelib/fireshape/innerproduct.py:180: in get_mu
    fd.solve(a == b, mu, bcs = [bc_fix, bc_free])
/home/wechsung/bin/firedrake/src/firedrake/firedrake/solving.py:121: in solve
    _solve_varproblem(*args, **kwargs)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/solving.py:150: in _solve_varproblem
    appctx=appctx)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/variational_solver.py:307: in __init__
    super(LinearVariationalSolver, self).__init__(*args, **kwargs)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/variational_solver.py:158: in __init__
    options_prefix=self.options_prefix)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/solving_utils.py:306: in __init__
    mat_type=mat_type)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/assemble.py:145: in create_assembly_callable
    collect_loops=True)
<decorator-gen-529>:2: in _assemble
    ???
/home/wechsung/bin/firedrake/src/firedrake/firedrake/utils.py:61: in wrapper
    return f(*args, **kwargs)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/assemble.py:569: in _assemble
    thunk(bcs)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/assemble.py:491: in thunk
    i, j)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/assemble.py:331: in mat
    m = testmap(test.function_space()[i])
/home/wechsung/bin/firedrake/src/firedrake/firedrake/assemble.py:489: in <lambda>
    tensor_arg = mat(lambda s: get_map(s, tsbc, decoration),
/home/wechsung/bin/firedrake/src/firedrake/firedrake/assemble.py:446: in get_map
    return x.cell_node_map(bcs)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/functionspaceimpl.py:451: in cell_node_map
    parent)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/functionspacedata.py:505: in get_map
    nodes = [bc.nodes for bc in explicit_bcs]
/home/wechsung/bin/firedrake/src/firedrake/firedrake/functionspacedata.py:505: in <listcomp>
    nodes = [bc.nodes for bc in explicit_bcs]
/home/wechsung/bin/firedrake/src/PyOP2/pyop2/utils.py:62: in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/bcs.py:179: in nodes
    return self._function_space.boundary_nodes(self.sub_domain, self.method)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/functionspaceimpl.py:514: in boundary_nodes
    return self._shared_data.boundary_nodes(self, sub_domain, method)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/functionspacedata.py:442: in boundary_nodes
    return get_boundary_nodes(V.mesh(), key, V)
<decorator-gen-517>:2: in get_boundary_nodes
    ???
/home/wechsung/bin/firedrake/src/firedrake/firedrake/functionspacedata.py:53: in cached
    result = f(mesh, key, *args, **kwargs)
/home/wechsung/bin/firedrake/src/firedrake/firedrake/functionspacedata.py:306: in get_boundary_nodes
    indices = dmplex.boundary_nodes(V, sub_domain, method)
firedrake/dmplex.pyx:703: in firedrake.dmplex.boundary_nodes
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <firedrake.mesh._Facets object at 0x7fe1c7cb4da0>, markers = ()

    def subset(self, markers):
        """Return the subset corresponding to a given marker value.
    
            :param markers: integer marker id or an iterable of marker ids
                (or ``None``, for an empty subset).
            """
        valid_markers = set([unmarked]).union(self.unique_markers)
        markers = as_tuple(markers, int)
        if self.markers is None and valid_markers.intersection(markers):
            return self._null_subset
        try:
            return self._subsets[markers]
        except KeyError:
            # check that the given markers are valid
            if len(set(markers).difference(valid_markers)) > 0:
                invalid = set(markers).difference(valid_markers)
                raise LookupError("{0} are not a valid markers (not in {1})".format(invalid, self.unique_markers))
    
            # build a list of indices corresponding to the subsets selected by
            # markers
            indices = np.concatenate([np.nonzero(self.markers == i)[0]
>                                     for i in markers])
E           ValueError: need at least one array to concatenate

/home/wechsung/bin/firedrake/src/firedrake/firedrake/mesh.py:188: ValueError
```


We can probably wait for the other bspline issues to be fixed until we attack this.

--- a/examples/levelset/levelset_spline.py
+++ b/examples/levelset/levelset_spline.py
@@ -9,7 +9,8 @@ bbox = [(-3.3, 4.3), (-3.3, 4.3)]
 orders = [3, 3]
 levels = [2, 2]
 Q = fs.BsplineControlSpace(mesh, bbox, orders, levels, fixed_dims=[0])
-inner = fs.H1InnerProduct(Q, fixed_bids=[1, 2, 3, 4])
+# inner = fs.H1InnerProduct(Q, fixed_bids=[1, 2, 3, 4])
+inner = fs.ElasticityInnerProduct(Q, fixed_bids=[1, 2, 3, 4])
 q = fs.ControlVector(Q, inner)
 
 mesh_m = Q.mesh_m

--- a/fireshape/innerproduct.py
+++ b/fireshape/innerproduct.py
@@ -210,14 +210,19 @@ class ElasticityInnerProduct(UflInnerProduct):
     """Inner product stemming from the linear elasticity equation."""
     def get_mu(self, V):
         W = fd.FunctionSpace(V.mesh(), "CG", 1)
-        bc_fix = fd.DirichletBC(W, 1, self.fixed_bids)
-        bc_free = fd.DirichletBC(W, 10, self.free_bids)
+        bcs = []
+        if len(self.fixed_bids):
+            bcs.append(fd.DirichletBC(W, 1, self.fixed_bids))
+        if len(self.free_bids):
+            bcs.append(fd.DirichletBC(W, 10, self.free_bids))
+        if len(bcs) == 0:
+            bcs = None
         u = fd.TrialFunction(W)
         v = fd.TestFunction(W)
         a = fd.inner(fd.grad(u), fd.grad(v)) * fd.dx
         b = fd.inner(fd.Constant(0.), v) * fd.dx
         mu = fd.Function(W)
-        fd.solve(a == b, mu, bcs=[bc_fix, bc_free])
+        fd.solve(a == b, mu, bcs=bcs)
         return mu
 
     def get_weak_form(self, V):

